### ISSUE TITLE
Generation FID is much lower than Reconstruction FID for models using VQ-16 (discrete) provided by LDM codebase

### ISSUE THREAD

**Issue by ShiFengyuan1999** [LTH14/mar#6]
Hi @LTH14, rFID of the tokenizer is usually the upper bound of the subsequent generation, but I find that transformers trained with VQ-16 (**5.15 rFID**) provided by LDM (https://github.com/CompVis/latent-diffusion) and Cross Entropy get **much lower** generation FID (**4.92 FID for AR version and 3.50 FID for MAR version**) than reconstruction FID in Table 1 of the paper. 

Can you provide implementation details of VQ baselines and give some insights about why generation FID is much lower than reconstruction FID for VQ cases?

### MAR Generation Results:
![MAR FID](https://github.com/user-attachments/assets/915152a5-d6d3-4a93-9394-60e9fae29bce)

### LDM VQ-16 rFID:
<img width="848" alt="LDM VQ-16 rFID" src="https://github.com/user-attachments/assets/6f0cf655-22e1-42c4-a6a6-6e20041d25b1">

**Comment by LTH14** [comment]
It is very easy to modify the current codebase's VAE to VQ-16 -- you only need to change the VAE and the scaling factor. We also observe this phenomenon. As shown in the table, the performance w.o. CFG is worse than the rFID, while the performance with CFG can achieve a better performance. My understanding is, CFG has the ability to modify the probability manifold of the generative model, so it has the potential to achieve a better FID than rFID. A similar thing is observed in the VQGAN paper as shown in this table:
<img width="334" alt="image" src="https://github.com/user-attachments/assets/b011c23e-3f83-4b87-8ce0-a626ff8ec1c7">
where the rFID of the VQGAN is 
<img width="343" alt="image" src="https://github.com/user-attachments/assets/9fc660c5-4de4-4133-bb30-f1a4b80d61e5">

The "Classifier-based rejection sampling" method later evolves into "Classifier guidance", and eventually becomes "Classifier-free Guidance".

**Comment by ShiFengyuan1999** [comment]
Hi @LTH14. Thanks for your response. The generation results of Table 4 in VQGAN paper are based on the **codebook of size 16384, and rFID should be 4.98**. So the **generation FID (5.20**) is still higher than the **reconstruction FID (4.98)**. 

<img width="512" alt="VQGAN FID" src="https://github.com/user-attachments/assets/ada86fb3-8f64-46f7-9764-1a65ac119c0e">

<img width="518" alt="VQGAN rFID" src="https://github.com/user-attachments/assets/13d961fc-e818-4600-89d1-2911e6ee99a5">


I agree that CFG or per-step rejection sampling can modify the probability manifold of the generative model and improve the generation performance. However, transformers are trained to generate an index sequence, and the evaluation is done on the images recovered by the decoder of VQGAN. If the index sequence is out of domain of VQGAN, the decoded images may be low-quality, and the FID will be inferior. It is somewhat surprising that generation FID is lower than reconstruction FID showing in your paper.

Could you provide the VQGANs (VQ-16) and transformers (Cross Entropy Loss) checkpoints achieving the results of the paper?

**Comment by LTH14** [comment]
Thanks for the clarification! But the 5.20 ImageNet FID is evaluated vs. ImageNet train instead of vs. ImageNet val, as shown in the table of the VQGAN GitHub repo: 

<img width="759" alt="image" src="https://github.com/user-attachments/assets/a9927655-7cd0-49f8-9084-468dcc714071">

The VQGAN rFID score vs. train is 7.41, while the generation FID score vs. train is 5.20, which is consistent with our observation. Additionally, it's important to note that the generation Inception Score (IS) can also significantly outperform the reconstruction Inception Score. It can even exceed the IS of ground-truth images. For example, setting DiT's cfg to 3.0 can result in an IS surpassing 400, whereas randomly sampling 50,000 ground-truth images from the ImageNet training set typically results in an IS below 350. The "mixed k, p = 1.0, acceptance rate 0.005" configuration in VQGAN also results in an IS of 402.7. My reasoning behind this observation is that the VQ reconstruction of some "easy" and simple images can be good, while for more "difficult" and complex images, it can be poor. CFG might force the generative model to produce only those "easy" examples. Although this approach may slightly reduce diversity, it can significantly enhance the average generation quality, thereby resulting in better FID and IS scores.

The VQGAN we use for the cross entropy loss is the same as [this one](https://heibox.uni-heidelberg.de/f/0e42b04e2e904890a9b6/?dl=1). [Here](https://drive.google.com/file/d/11m4Xf65-Cv0CeyYkMYOeLg9Qii5ofGfx/view?usp=sharing) is the MAR+CrossEntropy transformer checkpoint from our original implementation. The 3.69 FID is achieved with linear schedule CFG=4.0 and logits temperature 1.1. Since this repo is for MAR+DiffLoss we will not support the training/inference of Cross-entropy-based MAR in this repo. You can find a very similar implementation of VQ-based MAR in the [MAGE](https://github.com/LTH14/mage) repo.

**Comment by ShiFengyuan1999** [comment]
Thanks for your reply.
